### ISSUE TITLE
Layout is created on `makeUIViewController` only, therefore custom layouts are not really updated

### ISSUE THREAD

**Issue by nferruzzi** [apptekstudios/ASCollectionView#46]
Hi, 

I'm back with the same story but in a different sauce and still talking about the `AdjustableGridScreen`.
A few hours ago I understood that my columns were not changing due to 2 distinct problems, the first one is solved and the second one is what I'm writing about now.

First a short recap:
my observable object state was living within the screen itself, therefore losing all its custom values every time the parent view was recreating the screen itself (my fault). 
In my demo project that happened frequently due to a global @EnvironmentObject with other models triggering the refresh. 

Moving the state outside the view fixed this and with the layout closure capturing that global state the layout was properly refreshed. 

But that was regardless of `shouldInvalidateLayoutOnStateChange`, the closure is called but the `collectionView` layout is still the one created from `makeUIViewController`, and this is the reason variable wrapped with @State are seen as the "old" ones because they are captured by the closure as values.

To make this case working I suggest a change like this in `updateUIViewController` to really recreate the layout.

```
public func updateUIViewController(_ collectionViewController: AS_CollectionViewController, context: Context)
{
    context.coordinator.parent = self
    updateCollectionViewSettings(collectionViewController.collectionView, delegate: context.coordinator.delegate)
    if context.coordinator.parent.shouldInvalidateLayoutOnStateChange {
        debugPrint("Really recreate the layout")
        collectionViewController.collectionViewLayout = layout.makeLayout(withCoordinator: context.coordinator)
        collectionViewController.collectionView.collectionViewLayout = collectionViewController.collectionViewLayout
    }
    context.coordinator.updateLayout()
    context.coordinator.updateContent(collectionViewController.collectionView, animated: true, refreshExistingCells: true)
}

``` 

and eventually, add another modifier like `shouldRecreateLayoutOnStateChange`.

thank you!

**Comment by apptekstudios** [comment]
It's funny you suggest `shouldRecreateLayoutOnStateChange` because I'd already made that exact thing on my local branch! I've now pushed that to master.

However, I'd recommend **against** using that wherever possible, because UICollectionView tends to handle layout object changes quite poorly, therefore:

To make it super easy to update a custom layout, I've added a configuration handler that will be called with each state update. See the waterfall layout for a demo 👍 

As simple as:
```swift
ASCollectionView(...) { ... }
.layout(createCustomLayout: ASWaterfallLayout.init) { layout in
			//This closure is automatically called to update the layout
			layout.someLayoutSetting = self.theSetting
}
```
